kernel/queue.c

```diff
@@ -205,21 +205,25 @@ static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 	k_poll_event_init(&event, K_POLL_TYPE_FIFO_DATA_AVAILABLE,
 			  K_POLL_MODE_NOTIFY_ONLY, queue);
 
-	event.state = K_POLL_STATE_NOT_READY;
+	do {
+		event.state = K_POLL_STATE_NOT_READY;
 
-	err = k_poll(&event, 1, timeout);
-	if (err) {
-		return NULL;
-	}
+		err = k_poll(&event, 1, timeout);
+		if (err) {
+			return NULL;
+		}
 
-	__ASSERT_NO_MSG(event.state == K_POLL_STATE_FIFO_DATA_AVAILABLE);
+		__ASSERT_NO_MSG(event.state ==
+				K_POLL_STATE_FIFO_DATA_AVAILABLE);
+
+		/* sys_slist_* aren't threadsafe, so must be always protected by
+		 * irq_lock.
+		 */
+		key = irq_lock();
+		val = sys_slist_get(&queue->data_q);
+		irq_unlock(key);
+	} while (!val && timeout == K_FOREVER);
```

Comment:
Can we have a if-clause outside loop to check timeout == K_FOREVER? This is to avoid branching in loop. Lesser CPU time.

Reply:
Hmm where would you put it though? Btw, are you sure this would be optimized away by the compiler?

Reply:
@Vudentz I think @cvinayak means an early check to iterate to avoid having to test for K_FOREVER in every iteration of the loop

Reply:
@Vudentz my suggested diff to current code in this PR: https://gist.github.com/cvinayak/9bb2e7dc5da943df926cf48c4df3d169

or plain code:
https://gist.github.com/cvinayak/5f5dfed8b018038f98c34e7f6b2f68c8

Reply:
@cvinayak Since the loop does actually contain a call that can sleep/pend I don't think we need such optimization since it is very unlikely we will get many NULLs in a row to cause a substantial difference.

Reply:
@Vudentz its hard for me to see a while-condition with a variable that does not change inside the loop. There is always the possibility with the correct frequency of IRQ this loop can run for many iterations.

Reply:
FWIW: the compiler will surely hoist that "timeout == K_FOREVER" out of the loop.  I guess we could check the assembly to be sure, but I don't see anything with this as written.  It's clear and relying on a compiler optimizer is something we always do anyway.